Add taskProgressBar and options

eschramm opened the pull request:
Closes https://github.com/CareEvolution/RKStudio-Participant/issues/789.

**ORK1Kit (RK1)**

For RK1, it was pretty straightforward, except that in order to have control over sizing the `UIProgressView`, I needed to embed into a container view. This is using the standard thickness to match platform UI (which will increase some in iOS 14).

|Text|Bar|None|
|---|---|---|
|![1text](https://user-images.githubusercontent.com/1008462/90050058-4db57280-dc9b-11ea-8994-ed80588fe291.png)|![1bar](https://user-images.githubusercontent.com/1008462/90050054-4c844580-dc9b-11ea-9fb9-964a7bb9647a.png)|![1none](https://user-images.githubusercontent.com/1008462/90050061-4e4e0900-dc9b-11ea-953e-6d63173fda25.png)

**ResearchKit (RK2)**

In RK2, the standard behavior for displaying progress in text is to stuff it into a `UIBarButtonItem` due to moving the `title` into the navigation bar. Because of sizing, a `UIProgressView` is IMO too small to be usable there. Instead, this implementation places it into the RK2-specific navigation container which hovers from the bottom of the screen. Doing this requires some funky work to to connect the `ORKTaskViewController` to the `UIProgressView`. Due to order of operations in view life cycle, traversing the view hierarchy to find it doesn't work since at the point the block is fired, the view isn't entirely attached. Furthermore, because of updating the view after it appears, the changes to the `UIProgressView` animate strangely between steps. So ideally, when the `ORKNavigationContainerView` is created, it would know what the last step's state was as far as showing or hiding the `UIProgressView`. This necessitated a singleton to store that state to minimize non-sensical animation.

|Text|Bar|None|
|---|---|---|
|![2text](https://user-images.githubusercontent.com/1008462/90050065-4ee69f80-dc9b-11ea-8235-a514ad827849.png)|![2bar](https://user-images.githubusercontent.com/1008462/90050063-4ee69f80-dc9b-11ea-8b98-ffd79fbc3684.png)|![2none](https://user-images.githubusercontent.com/1008462/90050066-4f7f3600-dc9b-11ea-8060-af67028566c8.png)

eschramm committed:
add taskProgressBar and options

mmertsock commented:
> This necessitated a singleton to store that state to minimize non-sensical animation.

Is this the first use of a singleton for UI behavior in ResearchKit?

mmertsock commented on ORK1Kit/ORK1Kit/Common/CEVRK1NavigationBarProgressView.m:
```diff
@@ -0,0 +1,52 @@
+- (instancetype)initWithFrame:(CGRect)frame {
+    if (self = [super initWithFrame:frame]) {
+        [self createConstraints];
+        self.accessibilityLabel = @"NavigationBarProgressView";
```
Not a user-friendly string/needs localization

eschramm replied:
I actually was able to remove these - they were remnants from creating the UITests. It turns out, putting views into the nav bar prevents them from being accessible ☹️

mmertsock commented on ORK1Kit/ORK1Kit/Common/CEVRK1NavigationBarProgressView.m:
```diff
@@ -0,0 +1,52 @@
+- (void)createConstraints {
+    _progressView = [[UIProgressView alloc] initWithFrame:CGRectZero];
+    _progressView.progress = 0;
+    _progressView.accessibilityLabel = @"TaskProgressView";
```
Not a user-friendly string/needs localization

eschramm replied:
Same as above

mmertsock commented on ORK1Kit/ORK1Kit/Common/CEVRK1NavigationBarProgressView.m:
```diff
@@ -0,0 +1,52 @@
+    
+    // This forces the bar to stretch so the ORK1ProgressView will attempt to take up the entire available width
+    NSLayoutConstraint *widthConstraint = [self.widthAnchor constraintEqualToConstant:200];
+    widthConstraint.priority = NSURLSessionTaskPriorityHigh;
```
Use UILayoutPriorityDefaultHigh instead?

eschramm replied:
Ah, good catch, thank you, really miss Swift for these things.

mmertsock commented on ORK1Kit/ORK1Kit/Common/ORK1TaskViewController.m:
```diff
@@ -966,11 +976,14 @@ - (void)showViewController:(ORK1StepViewController *)viewController goForward:(B
-    if ([self shouldDisplayProgressLabel]) {
+    ORK1TaskProgress taskProgress;
+    taskProgress.current = 0;
+    taskProgress.total =  0;
```
These 3 lines could be shortened to `ORK1TaskProgress taskProgress = ORK1TaskProgressMake(0, 0);`

mmertsock commented on ORK1Kit/ORK1Kit/Common/ORK1TaskViewController.m:
```diff
@@ -1491,7 +1538,7 @@ - (void)decodeRestorableStateWithCoder:(NSCoder *)coder {
         
         if ([_task respondsToSelector:@selector(stepWithIdentifier:)]) {
-            _hasSetProgressLabel = [coder decodeBoolForKey:_ORK1HasSetProgressLabelRestoreKey];
+            _hasSetProgress = [coder decodeBoolForKey:_ORK1HasSetProgressRestoreKey];
```
Changing the restoration key may break the progress UI logic when restoring a task across ResearchKit version changes.

eschramm replied:
Good point, will revert.

mmertsock commented on ResearchKit/Common/ORKNavigationContainerView.m:
```diff
@@ -263,6 +265,12 @@ - (void)setupParentStackView {
+    self.taskProgressView = progressView;
+    [_parentStackView addArrangedSubview:progressView];
+    [progressView.topAnchor constraintEqualToAnchor:_parentStackView.topAnchor constant:10].active = YES;
+    [_parentStackView setCustomSpacing:20 afterView:progressView];
```
I didn't know about setCustomSpacing. Good to know!

mmertsock commented on ResearchKit/Common/ORKStepViewController.h:
```diff
@@ -357,6 +358,8 @@ ORK_CLASS_AVAILABLE
  */
 @property (nonatomic, weak, readonly, nullable) ORKTaskViewController *taskViewController;
 
+@property (nonatomic, weak, nullable) ORKNavigationContainerView *navigationContainerView;
```
Would it follow ResearchKit idioms better to declare this in ORKStepViewController\_Internal.h?

eschramm replied:
Good suggestion, done.

mmertsock reviewed (commented):
Code looks good overall. The core logic for progress calculation looks good to me, I thought through several use cases and it should be solid and match the web implementation.

eschramm committed:
remove singleton for RK2

eschramm committed:
PR fixes

- revert rename of _hasSetProgressLabel to omit Label
- use ORK(1)TaskProgressMake function
- move new properties to _Internal headers
- remove accessibility labels that are no longer needed, test-building artifact
- fix erroneous UILayoutPriority enum

eschramm committed:
fix landscape progressBar appearance

eschramm commented:
Was able to remove the singleton in RK2 by passing a reference to the StepViewController which can access the TaskViewController where we can store the last hidden state for the progress bar. 

I also found that simply embedding the `_parentStackView` inside another (`_grandparentStackView`) in the `ORKNavigationContainerView` would fix the wonky squishing/stretching seen in RK2 if we rotate to landscape mode.

BEFORE
![Simulator Screen Shot - iPhone 11 - 2020-08-14 at 11 07 23](https://user-images.githubusercontent.com/1008462/90271015-c5f97080-de20-11ea-92f7-ad6455b428db.png)


AFTER
![Simulator Screen Shot - iPhone 11 - 2020-08-14 at 11 13 37](https://user-images.githubusercontent.com/1008462/90271045-d14c9c00-de20-11ea-90b0-386020d975ee.png)

eschramm committed:
PR fix - missed RK1 for revert rename of _hasSetProgressLabel to omit Label

mmertsock commented on ResearchKit/Common/ORKStepViewController.h:
```diff
@@ -43,6 +43,7 @@ NS_ASSUME_NONNULL_BEGIN
 @class ORKStepResult;
 @class ORKStepViewController;
 @class ORKTaskViewController;
+@class ORKNavigationContainerView;
```
This forward declaration should be moved to ORKSteViewController\_Internal.h

mmertsock reviewed (approved):
LGTM 🚢

eschramm committed:
fix for ORKConsentSceneViewController to fix failing test

eschramm committed:
move forward declaration to internal header
